mempool/src/shared_mempool/use_case_history.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_mempool_notifications::CommittedTransaction;
+use aptos_types::transaction::use_case::UseCaseKey;
+use std::collections::{HashMap, HashSet, VecDeque};
+
+pub(crate) struct UseCaseHistory {
+    window_size: usize,
+    num_top_to_track: usize,
+    recent: VecDeque<HashMap<UseCaseKey, usize>>,
+}
+
+impl UseCaseHistory {
+    pub(crate) fn new(window_size: usize, num_top_to_track: usize) -> Self {
+        Self {
+            window_size,
+            num_top_to_track,
+            recent: VecDeque::with_capacity(window_size + 1),
+        }
+    }
+
+    pub(crate) fn update_usecases_and_get_tracking_set(
+        &mut self,
+        transactions: &[CommittedTransaction],
+    ) -> HashSet<UseCaseKey> {
+        let mut count_by_usecase = HashMap::new();
+        for transaction in transactions {
+            *count_by_usecase
+                .entry(transaction.use_case.clone())
+                .or_insert(0) += 1;
+        }
+
+        self.recent.push_back(count_by_usecase);
+        while self.recent.len() > self.window_size {
+            self.recent.pop_front();
+        }
+
+        let mut total = HashMap::new();
+        for group in &self.recent {
+            for (use_case, count) in group {
+                if use_case != &UseCaseKey::Platform && use_case != &UseCaseKey::Others {
+                    *total.entry(use_case.clone()).or_insert(0) += count;
+                }
+            }
+        }
+
+        let mut result = HashSet::new();
+        result.insert(UseCaseKey::Platform);
+        result.insert(UseCaseKey::Others);
+
+        let mut sorted = total.into_iter().collect::<Vec<_>>();
+        sorted.sort_by_key(|(_, count)| *count);
+
+        for _ in 0..self.num_top_to_track {
+            if let Some((use_case, _)) = sorted.pop() {
+                result.insert(use_case);
+            }
+        }
+
+        result
```

Comment:
probably doesn't matter, but perfect use case for a `BinaryHeap<(usize, UseCaseKey)>` 😂

Reply:
you mean to use heap instead of total, or just instead of sorted? 

if instead of total, not sure how to update point values for usecase.

if instead of sorting - does rust have a way to convert vector to heap in O(n)? If inserting from iterator - complexity will be the same.

Reply:
I mean sorting. 
true it's the same O, but on average it must be faster than completely sorting them, right? 
https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html#time-complexity-3

but again, this probably matters too little even for me

Reply:
heap can be created in O(n), worst case, but I don't see rust providing such API?

Reply:
I think even pushing one by one is probably meaningfully faster -- as it grows most likely a new item doesn't need to climb many levels of the heap -- that's why they estimate ~O(1) for pushing as well

Reply:
that's average. Reverse sorted would be O(n log n). but yeah, I can switch to heap, unfortunate I cannot get O(n)

Reply:
actually they do have that optimization..
 https://github.com/rust-lang/rust/blob/fcc325f1bc477975e2ce5ba534fe4c77ff8a8536/library/alloc/src/collections/binary_heap/mod.rs#L1809